Sources/SourceControl/RepositoryManager.swift

```diff
@@ -164,55 +159,120 @@ public class RepositoryManager {
 
     /// Get a handle to a repository.
     ///
-    /// This will initiate a clone of the repository automatically, if
-    /// necessary, and immediately return. The client can add observers to the
-    /// result in order to know when the repository is available.
-    public func lookup(repository: RepositorySpecifier) -> RepositoryHandle {
-        // Check to see if the repository has been provided.
-        if let handle = repositories[repository.url] {
-            // Fetch and update the repository when it is being looked up.
-            // We need some mechanism to maybe not do this for pinned repos.
-            do {
-                let repo = try handle.open()
-                try repo.fetch()
-            } catch {
-                // FIXME: Better error handling.
-                print("Error while fetching the repo: \(handle)")
+    /// This will initiate a clone of the repository automatically, if necessary.
+    public func lookup(repository: RepositorySpecifier, completion: @escaping LookupCompletion) {
+        concurrentQueue.async {
+            // Get the handle and status. The completion will be stored or chained inside handle.
+            let (handle, status) = self.getHandle(repository: repository, lookupCompletion: completion)
+            let result: LookupResult
```

Comment:
This code is still unprotected, and can race with other clients (e.g., two clients could trigger a fetch simultaneously.